test/unit/org/apache/cassandra/db/guardrails/GuardrailMaximumReplicationFactorTest.java

```diff
@@ -0,0 +1,246 @@
+        guardrails().setMaximumReplicationFactorThreshold(MAXIMUM_REPLICATION_FACTOR_WARN_THRESHOLD, MAXIMUM_REPLICATION_FACTOR_FAIL_THRESHOLD);
+        assertValid("CREATE KEYSPACE ks WITH replication = { 'class' : 'NetworkTopologyStrategy', 'datacenter1': 2, 'datacenter2' : 2 };");
+        execute("DROP KEYSPACE IF EXISTS ks");
+        assertWarns("CREATE KEYSPACE ks WITH replication = { 'class' : 'NetworkTopologyStrategy', 'datacenter1': 2, 'datacenter2' : 3 };",
+                    format("The keyspace ks has a replication factor of 3, above the warning threshold of %s.", MAXIMUM_REPLICATION_FACTOR_WARN_THRESHOLD));
+        execute("DROP KEYSPACE IF EXISTS ks");
+        assertWarns("CREATE KEYSPACE ks WITH replication = { 'class' : 'NetworkTopologyStrategy', 'datacenter1': 3, 'datacenter2' : 3 };", twoWarnings);
+        execute("DROP KEYSPACE IF EXISTS ks");
+        assertFails("CREATE KEYSPACE ks WITH replication = { 'class' : 'NetworkTopologyStrategy', 'datacenter1': 2, 'datacenter2' : 5 };",
+                    format("The keyspace ks has a replication factor of 5, above the failure threshold of %s", MAXIMUM_REPLICATION_FACTOR_FAIL_THRESHOLD));
+        execute("DROP KEYSPACE IF EXISTS ks");
+        assertFails("CREATE KEYSPACE ks WITH replication = { 'class' : 'NetworkTopologyStrategy', 'datacenter1': 5, 'datacenter2' : 5 };",
+                    format("The keyspace ks has a replication factor of 5, above the failure threshold of %s", MAXIMUM_REPLICATION_FACTOR_FAIL_THRESHOLD));
+        execute("DROP KEYSPACE IF EXISTS ks");
+
+        execute("CREATE KEYSPACE ks WITH replication = { 'class' : 'NetworkTopologyStrategy', 'datacenter1': 1, 'datacenter2' : 1 };");
+        assertValid("ALTER KEYSPACE ks WITH replication = { 'class' : 'NetworkTopologyStrategy', 'datacenter1': 2, 'datacenter2' : 2 };");
+        assertWarns("ALTER KEYSPACE ks WITH replication = { 'class' : 'NetworkTopologyStrategy', 'datacenter1': 2, 'datacenter2' : 3 };",
+                    format("The keyspace ks has a replication factor of 3, above the warning threshold of %s.", MAXIMUM_REPLICATION_FACTOR_WARN_THRESHOLD));
+        assertWarns("ALTER KEYSPACE ks WITH replication = { 'class' : 'NetworkTopologyStrategy', 'datacenter1': 3, 'datacenter2' : 3 };", twoWarnings);
+        assertFails("ALTER KEYSPACE ks WITH replication = { 'class' : 'NetworkTopologyStrategy', 'datacenter1': 2, 'datacenter2' : 5 };",
+                    format("The keyspace ks has a replication factor of 5, above the failure threshold of %s", MAXIMUM_REPLICATION_FACTOR_FAIL_THRESHOLD));
+        assertFails("ALTER KEYSPACE ks WITH replication = { 'class' : 'NetworkTopologyStrategy', 'datacenter1': 5, 'datacenter2' : 5 };",
+                    format("The keyspace ks has a replication factor of 5, above the failure threshold of %s", MAXIMUM_REPLICATION_FACTOR_FAIL_THRESHOLD));
```

Comment:
The phrases of the form `The keyspace ks has a replication factor...` are repeated multiple times across the test. We could try to reduce duplication and make the tests a bit more compact with a couple of utility methods like, for example:
```java
private void assertWarns(String query, int rf) throws Throwable
{
    assertWarns(query, format("The keyspace ks has a replication factor of %d, above the warning threshold of %s.",
                              rf, MAXIMUM_REPLICATION_FACTOR_WARN_THRESHOLD));
}

private void assertFails(String query, int rf) throws Throwable
{
    assertFails(query, format("The keyspace ks has a replication factor of %d, above the failure threshold of %s.",
                              rf, MAXIMUM_REPLICATION_FACTOR_FAIL_THRESHOLD));
}
```